### PR TITLE
concretely type more things

### PR DESCRIPTION
**Pull request by vchuravy**
addresses most of #251

### CODE REVIEW

**Comment by glwagner** [comment]
Nice. It's just `ModelBoundaryConditions` and `StepperTemporaryFields` remaining.

I can do the boundary conditions --- though it changes the user API, because the boundary conditions must be specified in the model constructor and their type cannot be changed afterwards (usually this is not a huge restriction).

**Comment by vchuravy** [comment]
All of this is based on specifing the types at model constructions.

**Comment by glwagner** [comment]
Right, and I understand now there's no free lunch on this subject!

I think there is some API design that will need to be done to make specifying boundary conditions simple once the model is completely statically typed (its complicated because there are potentially 30 boundary conditions that can be set on 5 3D fields), but this is future work.

**Comment by glwagner** [comment]
I'm not sure how I marked this ready for review but I didn't actually mean to. Though it does seem ready.

**Comment by glwagner** [comment]
Well, I made a bunch of changes, but either I'm totally bungling this dance-of-the-git, or I don't have permission to push changes to this PR (probably the former)...

**Comment by vchuravy** [comment]
This is on my own fork so you just don't have permissions.

File: src/models.jl [hunk 1]
```diff
@@ -1,21 +1,24 @@
 using .TurbulenceClosures
 
-mutable struct Model{A<:Architecture, G, TC, T}
+mutable struct Model{A<:Architecture, Grid, TC, T, F,
+                    PC<:PlanetaryConstants, PS, VC<:VelocityFields,
+                    EOS<:EquationOfState, TG, TGp,
+                    Tracers<:TracerFields, PF<:PressureFields}
```
Comment (ali-ramadhan) [inline src/models.jl]: 🚨 Violation of Julia naming convention? Doesn't matter to me, we can fix later.

**Comment by ali-ramadhan** [comment]
Thanks for doing this! Stacktraces are much longer now lol but might help with abstraction performance?

This is not ready to be merged though. The tests are failing because of some error when assigning the forcing functions in the `Model` constructor. I can have a closer look and figure out a fix.

> Nice. It's just `ModelBoundaryConditions` and `StepperTemporaryFields` remaining.

It's possible that `StepperTemporaryFields` isn't even required anymore. If so, I'll open a new PR removing them.

**Comment by glwagner** [comment]
I fixed those problems on my branch:

https://github.com/glwagner/Oceananigans.jl/tree/concrete

Not sure what's best way to proceed.

**Comment by vchuravy** [comment]
just open a PR of that branch

**Comment by glwagner** [comment]
Thanks @vchuravy !

**Pull request closed without merging**